File: flow_log_processor.py
#!/usr/bin/env python3

import sys
import csv
from collections import defaultdict
from typing import Dict, Tuple, List

def load_lookup_table(lookup_file: str) -> Dict[Tuple[str, str], str]:
    """
    Load the lookup table from CSV file and create a mapping of (port, protocol) to tag.
    """
    lookup = {}
    with open(lookup_file, 'r') as f:
        reader = csv.DictReader(f)
        for row in reader:
            port = row['dstport']
            protocol = row['protocol'].lower()  # Convert to lowercase for case-insensitive matching
            tag = row['tag']
            lookup[(port, protocol)] = tag
    return lookup

def process_flow_log(flow_file: str, lookup: Dict[Tuple[str, str], str]) -> Tuple[Dict[str, int], Dict[Tuple[str, str], int]]:
    """
    Process the flow log file and count tags and port/protocol combinations.
    """
    tag_counts = defaultdict(int)
    port_protocol_counts = defaultdict(int)
    
    # Protocol mapping from number to name
    protocol_map = {
        '6': 'tcp',
        '17': 'udp'
    }
    
    with open(flow_file, 'r') as f:
        for line in f:
            fields = line.strip().split()
            if len(fields) < 8:  # Skip invalid lines
                continue
                
            # Extract relevant fields (based on AWS VPC flow log format)
            dst_port = fields[6]
            protocol_num = fields[7]
            
            if protocol_num not in protocol_map:
                continue
                
            protocol = protocol_map[protocol_num]
            port_protocol = (dst_port, protocol)
            port_protocol_counts[port_protocol] += 1
            
            # Look up tag
            tag = lookup.get(port_protocol, "Untagged")
            tag_counts[tag] += 1
    
    return tag_counts, port_protocol_counts

def print_results(tag_counts: Dict[str, int], port_protocol_counts: Dict[Tuple[str, str], int]):
    """
    Print the results in the required format.
    """
    # Print tag counts
    print("\nTag Counts:")
    print("Tag             Count")
    print("-" * 20)
    for tag in sorted(tag_counts.keys()):
        print(f"{tag:<15} {tag_counts[tag]}")
    
    # Print port/protocol counts
    print("\nPort/Protocol Combination Counts:")
    print("Port    Protocol        Count")
    print("-" * 30)
    for (port, protocol), count in sorted(port_protocol_counts.items()):
        print(f"{port:<8} {protocol:<15} {count}")

def main():
    if len(sys.argv) != 3:
        print("Usage: python flow_log_processor.py <flow_log_file> <lookup_table_file>")
        sys.exit(1)
    
    flow_file = sys.argv[1]
    lookup_file = sys.argv[2]
    
    try:
        lookup = load_lookup_table(lookup_file)
        tag_counts, port_protocol_counts = process_flow_log(flow_file, lookup)
        print_results(tag_counts, port_protocol_counts)
    except Exception as e:
        print(f"Error: {str(e)}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main() 

File: src/main/java/com/illumio/FlowLogProcessor.java
package com.illumio;

import com.illumio.model.LookupEntry;
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class FlowLogProcessor {
    // TCP = 6, UDP = 17 in VPC Flow Logs
    private static final Map<String, String> PROTOCOL_MAP = Map.of(
        "6", "tcp",
        "17", "udp"
    );
    
    private final Map<LookupEntry, String> lookupTable;
    private final Map<String, Integer> tagCounts;
    private final Map<LookupEntry, Integer> portProtocolCounts;

    public FlowLogProcessor(String lookupFile) throws IOException {
        if (!Files.exists(Paths.get(lookupFile))) {
            throw new FileNotFoundException(lookupFile);
        }
        this.lookupTable = loadLookupTable(lookupFile);
        this.tagCounts = new HashMap<>();
        this.portProtocolCounts = new HashMap<>();
    }

    private Map<LookupEntry, String> loadLookupTable(String lookupFile) throws IOException {
        Map<LookupEntry, String> lookup = new HashMap<>();
        try (BufferedReader reader = Files.newBufferedReader(Paths.get(lookupFile))) {
            String header = reader.readLine();
            if (!header.equals("dstport,protocol,tag")) {
                throw new IllegalArgumentException("Invalid lookup table format");
            }

            String line;
            while ((line = reader.readLine()) != null) {
                if (line.trim().isEmpty()) continue;
                String[] parts = line.split(",");
                if (parts.length != 3) {
                    System.err.println("Warning: Skipping invalid line: " + line);
                    continue;
                }
                
                try {
                    int port = Integer.parseInt(parts[0].trim());
                    if (port >= 0 && port <= 65535) {
                        LookupEntry entry = new LookupEntry(
                            parts[0].trim(), 
                            parts[1].trim(), 
                            parts[2].trim()
                        );
                        lookup.put(entry, parts[2].trim());
                    }
                } catch (NumberFormatException e) {
                    // Skip invalid port numbers
                }
            }
        }
        return lookup;
    }

    public void processFlowLog(String flowFile, String outputFile) throws IOException {
        if (!Files.exists(Paths.get(flowFile))) {
            throw new FileNotFoundException(flowFile);
        }

        // Validate output file path
        Path outputPath = Paths.get(outputFile);
        if (outputPath.getParent() != null && !Files.exists(outputPath.getParent())) {
            Files.createDirectories(outputPath.getParent());
        }
        if (Files.exists(outputPath) && !Files.isWritable(outputPath)) {
            throw new IOException("Output file exists and is not writable: " + outputFile);
        }

        try (BufferedReader reader = Files.newBufferedReader(Paths.get(flowFile))) {
            String line;
            while ((line = reader.readLine()) != null) {
                try {
                    processLine(line.trim());
                } catch (Exception e) {
                    // Skip invalid lines
                }
            }
        }

        // Write results to output file
        try (PrintWriter writer = new PrintWriter(Files.newBufferedWriter(Paths.get(outputFile)))) {
            // Write tag counts
            writer.println("\nTag Counts:\n");
            writer.println("Tag             Count");
            writer.println("--------------------");
            
            tagCounts.entrySet().stream()
                    .sorted(Map.Entry.comparingByKey())
                    .forEach(entry -> writer.printf("%-16s %d%n", entry.getKey(), entry.getValue()));

            // Write port/protocol combination counts
            writer.println("\nPort/Protocol Combination Counts:\n");
            writer.println("Port    Protocol        Count");
            writer.println("------------------------------");
            portProtocolCounts.entrySet().stream()
                .sorted((a, b) -> {
                    int portCompare = Integer.compare(
                        Integer.parseInt(a.getKey().getDstPort()), 
                        Integer.parseInt(b.getKey().getDstPort())
                    );
                    return portCompare != 0 ? portCompare : 
                           a.getKey().getProtocol().compareTo(b.getKey().getProtocol());
                })
                .forEach(entry -> {
                    writer.printf("%-8s %-15s %d%n", 
                        entry.getKey().getDstPort(), 
                        entry.getKey().getProtocol(), 
                        entry.getValue());
                });
        }
    }

    private void processLine(String line) {
        String[] fields = line.split("\\s+");
        // VPC Flow Log format needs at least these fields
        if (fields.length < 8) return;

        String dstPort = fields[6];
        String protocolNum = fields[7];
        
        try {
            int port = Integer.parseInt(dstPort);
            if (port < 0 || port > 65535) return;
            
            String protocol = PROTOCOL_MAP.get(protocolNum);
            if (protocol == null) return;

            LookupEntry entry = new LookupEntry(dstPort, protocol, null);
            portProtocolCounts.merge(entry, 1, Integer::sum);

            String tag = lookupTable.getOrDefault(entry, "Untagged");
            tagCounts.merge(tag, 1, Integer::sum);
        } catch (NumberFormatException e) {
            // Skip invalid port numbers
        }
    }

    // Getter methods for testing
    public Map<String, Integer> getTagCounts() {
        return Collections.unmodifiableMap(tagCounts);
    }

    public Map<LookupEntry, Integer> getPortProtocolCounts() {
        return Collections.unmodifiableMap(portProtocolCounts);
    }

    public static void main(String[] args) {
        if (args.length != 3) {
            System.err.println("Usage: java FlowLogProcessor <flow_log_file> <lookup_table_file> <output_file>");
            System.exit(1);
        }

        try {
            FlowLogProcessor processor = new FlowLogProcessor(args[1]);
            processor.processFlowLog(args[0], args[2]);
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }
} 